### DISCUSSION TITLE
Voice recognition does not work without https

### DISCUSSION THREAD

**Discussion by raldone01** [open-webui/open-webui#1148] [General]
# Bug Report

## Description

**Bug Summary:**
`Uncaught (in promise) TypeError: navigator.mediaDevices is undefined`
On firefox `navigator.mediaDevices` is only available in secure contexts (https).

**Steps to Reproduce:**
* Click on the microphone icon when connected to the webui via http.
* Check the developer console.

**Expected Behavior:**
* It should work or at least show some error outlining the https requirement and instruct users to setup a reverse proxy.

**Actual Behavior:**
* No error indication to the user.

## Environment

- **Operating System:** Arch
- **Browser (if applicable):** Firefox

Keep up the good work.

**Comment by UserName-wang** [comment]
my question is how to enable https connection?

**Comment by ghost** [comment]
I like the idea of notifying the user if `navigator.mediaDevices` is undefined but are you accessing Open Web-UI over localhost (as described in the [documentation](https://docs.openwebui.com/#quick-start-with-docker-))? This shouldn't be a problem because while the [Mozilla documentation states](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaDevices):
> _Secure context: This feature is available only in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (HTTPS), in some or all [supporting browsers](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaDevices#browser_compatibility)._

Communication with localhost is [considered to be a secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure):
> _Locally-delivered resources such as those with `http://127.0.0.1` URLs, `http://localhost` and `http://*.localhost` URLs (e.g. `http://dev.whatever.localhost/`), and `file://` URLs are also considered to have been delivered securely._

I cannot re-produce this with Arch Linux using the native Firefox package while connecting over localhost. A better overall solution may be to require [Strict-Transport-Security](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security) for non-local connections perhaps with an exclusion for proxied connections using [X-Forwarded-For](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) where the responsibility to set and manage HTST & TLS transport encryption can be presumed to occur there.

**Reply by raldone01** [reply]
I host it on my server and have yet to setup a https cert. I know how to do it but a notification would be nice so people know that they need https (or localhost) for that feature.

**Reply by slash-proc** [reply]
It's in the FAQ
https://docs.openwebui.com/faq#q-why-doesnt-stttts-work-in-my-deployment

**Comment by lee-b** [comment]
The proper solution is to add TLS (https) certificates, but as a quick workaround, for firefox, add the host to dom.securecontext.allowlist in about:config.  It's a string of comma-separated hosts.  Set it to 192.168.0.2, for example, if that's the IP of your computer and the IP that you're connecting to with Firefox.

**Comment by thiswillbeyourgithub** [comment]
Actually it is possible to add an exception to any URL but it's well hidden. I made a discussion post there but pinging didn't work so [linking here instead](https://github.com/open-webui/open-webui/discussions/4352)

This way I can use voice recognition / STT on all my devices including android
